compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1653,8 +1654,157 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 ),
                 Mismatch::Fixed(s) => (s.into(), s.into(), None),
             };
-            match (&terr, expected == found) {
-                (TypeError::Sorts(values), extra) => {
+
+            enum Similar<'tcx> {
+                Adts(ty::AdtDef<'tcx>, ty::AdtDef<'tcx>),
+                PrimitiveFound(Ty<'tcx>, ty::AdtDef<'tcx>),
+                PrimitiveExpected(ty::AdtDef<'tcx>, Ty<'tcx>),
+            }
+
+            let primitive_sym = |kind: &_| match kind {
+                ty::Bool => Some(sym::bool),
+                ty::Char => Some(sym::char),
+                ty::Float(f) => match f {
+                    ty::FloatTy::F32 => Some(sym::f32),
+                    ty::FloatTy::F64 => Some(sym::f64),
+                },
+                ty::Int(f) => match f {
+                    ty::IntTy::Isize => Some(sym::isize),
+                    ty::IntTy::I8 => Some(sym::i8),
+                    ty::IntTy::I16 => Some(sym::i16),
+                    ty::IntTy::I32 => Some(sym::i32),
+                    ty::IntTy::I64 => Some(sym::i64),
+                    ty::IntTy::I128 => Some(sym::i128),
+                },
+                ty::Uint(f) => match f {
+                    ty::UintTy::Usize => Some(sym::usize),
+                    ty::UintTy::U8 => Some(sym::u8),
+                    ty::UintTy::U16 => Some(sym::u16),
+                    ty::UintTy::U32 => Some(sym::u32),
+                    ty::UintTy::U64 => Some(sym::u64),
+                    ty::UintTy::U128 => Some(sym::u128),
+                },
+                _ => None,
+            };
+
+            let similarity = |e: ExpectedFound<Ty<'tcx>>| {
+                let (fk, ek) = (e.found.kind(), e.expected.kind());
+                match (fk, ek) {
+                    (
+                        ty::Adt(adt, _),
+                        ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::Float(_),
+                    ) => {
+                        let path = self.tcx.def_path(adt.did()).data;
+                        let name = path.last().unwrap().data.get_opt_name();
+                        let prim_sym = primitive_sym(ek);
+
+                        if name == prim_sym {
+                            return Some(Similar::PrimitiveExpected(*adt, e.expected));
+                        }
+                        None
+                    }
+                    (
+                        ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::Float(_),
```

Comment:
same above. I think you can also use an `if-let` guard to extract the primitive sym in this match statement directly.